coderd/database/dbtestutil/postgres.go

```diff
@@ -16,119 +25,452 @@ import (
+		// The fact that openContainer succeeded does not
+		// mean that port forwarding is ready.
+		for i := 0; i < 20; i++ {
+			db, err = sql.Open("postgres", dsn)
+			if err == nil {
+				err = db.Ping()
+				if closeErr := db.Close(); closeErr != nil {
+					return xerrors.Errorf("close db, container: %w", closeErr)
+				}
+			}
+			if err == nil {
+				break
+			}
+			time.Sleep(250 * time.Millisecond)
+		}
+	} else if err != nil {
+		return xerrors.Errorf("open postgres connection: %w", err)
+	}
+	defaultConnectionParams = params
+	return nil
+}
+
+type OpenOptions struct {
+	DBFrom *string
+}
+
+type OpenOption func(*OpenOptions)
+
+// WithDBFrom sets the template database to use when creating a new database.
+// Overrides the DB_FROM environment variable.
+func WithDBFrom(dbFrom string) OpenOption {
+	return func(o *OpenOptions) {
+		o.DBFrom = &dbFrom
+	}
+}
 
-		dbName, err := cryptorand.StringCharset(cryptorand.Lower, 10)
+// Open creates a new PostgreSQL database instance.
+// If there's a database running at localhost:5432, it will use that.
+// Otherwise, it will start a new postgres container.
+func Open(opts ...OpenOption) (string, func(), error) {
+	connectionParamsInitOnce.Do(func() {
+		errDefaultConnectionParamsInit = initDefaultConnectionParams()
+	})
+	if errDefaultConnectionParamsInit != nil {
+		return "", func() {}, xerrors.Errorf("init default connection params: %w", errDefaultConnectionParamsInit)
+	}
+
+	openOptions := OpenOptions{}
+	for _, opt := range opts {
+		opt(&openOptions)
+	}
+
+	var (
+		username = defaultConnectionParams.Username
+		password = defaultConnectionParams.Password
+		host     = defaultConnectionParams.Host
+		port     = defaultConnectionParams.Port
+	)
+
+	// Use a time-based prefix to make it easier to find the database
+	// when debugging.
+	now := time.Now().Format("test_2006_01_02_15_04_05")
+	dbSuffix, err := cryptorand.StringCharset(cryptorand.Lower, 10)
+	if err != nil {
+		return "", func() {}, xerrors.Errorf("generate db suffix: %w", err)
+	}
+	dbName := now + "_" + dbSuffix
+
+	// if empty createDatabaseFromTemplate will create a new template db
+	templateDBName := os.Getenv("DB_FROM")
+	if openOptions.DBFrom != nil {
+		templateDBName = *openOptions.DBFrom
+	}
+	if err = createDatabaseFromTemplate(defaultConnectionParams, dbName, templateDBName); err != nil {
+		return "", func() {}, xerrors.Errorf("create database: %w", err)
+	}
+
+	cleanup := func() {
+		cleanupDbURL := defaultConnectionParams.DSN()
+		cleanupConn, err := sql.Open("postgres", cleanupDbURL)
 		if err != nil {
-			return "", nil, xerrors.Errorf("generate db name: %w", err)
+			_, _ = fmt.Fprintf(os.Stderr, "cleanup database %q: failed to connect to postgres: %s\n", dbName, err.Error())
 		}
-
-		dbName = "ci" + dbName
-		_, err = db.Exec("CREATE DATABASE " + dbName + " WITH TEMPLATE " + os.Getenv("DB_FROM"))
+		defer cleanupConn.Close()
+		_, err = cleanupConn.Exec("DROP DATABASE " + dbName + ";")
```

Comment:
Won't these panic if `cleanupConn` is nil? The `err` check above does not exit the function

Reply:
good catch!